### PR TITLE
Basic testing workshop

### PR DESCRIPTION
**Pull request by njseeto**
Hi All 👋

Hope everyone is doing ok in these current covid times.

I thought I'd use the extra time to write up a simple server testing workshop.
It is written as an intro into testing endpoints and only looks at testing a `GET` and a `POST` as these are the 2 endpoints covered in the main Node workshop.

I have set it out so it is all encased in a `test` folder, and each step has its own readme.

I would love any feedback on the content (readability / level of information), also any other suggestions/additions/changes.

I don't have a solution branch but am happy to make one as a reference, if we think that is a good idea?

Stay safe, hope to see you all in the flesh soon!

### CODE REVIEW

File: test/readme.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Express Workshop - Testing Your Endpoints
+
+### What?
+You've probably finished the Node workshop and are now wondering how to test your endpoints. Firstly, well done on completing the workshop! Secondly, kudos for wanting to test those new endpoints you've built.
+
+### Why should we test?
+Testing helps you write better code. It's is a great way to make sure your code is doing what it is supposed to do.
+
+It also gives you the confidence to change your code with the knowledge that your tests will tell you if something has gone wrong.
```
Comment (minaorangina) [inline test/readme.md]: I really like the "why should we test" bit - testing as a concept can be hard for beginners to grasp. Some examples of what tests can catch would be good, e.g. if you put a typo in one of the endpoints, or accidentally use `POST` where you meant it to be `GET`. Could also mention how tests are way quicker/more accurate than checking things manually, especially if you have a lot of code.

File: test/tutorial/step-1-setting-up.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+### Setting Up
+First we need to install Jest, you can do this by running the following in your command line
+
+```
+npm install jest --save-dev
+```
+
+
+Next we want to configure the scripts in your `package.json` file.
```
Comment (minaorangina) [inline test/tutorial/step-1-setting-up.md]: Students won't have interacted with npm scripts yet at this stage, so this is a nice moment to discuss what `scripts` are generally for.  Maybe you could invite them to type `npm [run] test`, see what happens, then replace what's there with the `jest` command, to help with understanding.

I'll leave this up to you though, it might be too much of a detour from the actual testing.

File: test/tutorial/step-1-setting-up.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+```
+"scripts": {
+  "test": "jest"
+}
+```
+
+Now we need to create a file where we can write our tests. So go ahead and create a file called `server.test.js` in the root of the test folder.
```
Comment (minaorangina) [inline test/tutorial/step-1-setting-up.md]: Ask people to create the test folder explicitly, in case some panic that they can't find it. 
Could maybe say a word about test file naming conventions, i.e. explain why `server.test.js` has its name

File: test/tutorial/step-2-writing-your-first-test.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+Ok this is the exciting bit! We get to write our first test!!
+
+A few things before we get stuck in, Jest has 3 words that you can use:
+- `describe` : this lets you wrap a bunch of related tests together, it's used for organising test (we don't need to worry about it for now).
+- `it` : lets you run a test. We need this in every test!
+- `expect` : lets you write assertions. The tests pass if the assertions pass.
+
+Dont' worry this will all make sense when we start writing some tests.
+
+In your `server.test.js` file we can write an example of a failing test. Go ahead and write the following in your `server.test.js` file:
```
Comment (minaorangina) [inline test/tutorial/step-2-writing-your-first-test.md:11]: I think it's useful to explain why failing tests are a desirable thing .  It doesn't have to be a whole TDD 101, but I reckon it's useful to understand why writing failing tests is something you would want to do.

Maybe it could me mentioned later. I think it's good to not get in people's way with too much theory and actually let them write some code.  Do what you think is best :)

File: test/tutorial/step-3-setting-up-supertest.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+### Setting up Supertest
```
Comment (minaorangina) [inline test/tutorial/step-3-setting-up-supertest.md:1]: I'd reiterate what supertest is, in case people have forgotten

File: test/tutorial/step-3-setting-up-supertest.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+To use supertest we need to require supertest and app into your test file, to do this add this to the top of your `server.test.js` file:
+
+```
+const supertest = require('supertest')
+const app = require('../server')
+const request = supertest(app)
+```
+
+The `request` is what will be going to your server, we will be asserting on the response that comes back.
```
Comment (minaorangina) [inline test/tutorial/step-3-setting-up-supertest.md]: I'd reword "asserting" here, might not mean much to a beginner

File: test/tutorial/step-5-ports.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
+### Ports can only take one request at a time...
```
Comment (minaorangina) [inline test/tutorial/step-5-ports.md]: More accurately, ports can only have one server listening on them at a time

File: test/tutorial/step-6-asynchronous-testing.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    expect(response.status).toBe(200)
+
+    done()
+})
+```
+
+Let's break this down:
+ - The `async` function is always coupled with an `await` expression.
+ - The `async` function will wait until the `await` expression has completed, in our case the get request. Once that has completed the function will continue.
```
Comment (minaorangina) [inline test/tutorial/step-6-asynchronous-testing.md:39]: I'm wondering if it would be helpful to use the callback stuff in the main tutorial as a point of comparison, as both are ways are dealing with asynchronous code.  
No strong opinions though, it might just be overkill to go down that route 🤷

File: test/tutorial/step-7-testing-post-endpoints.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
+### Testing POST endpoints
+
+If we remember in our Node-Girls workshop, the `POST` request is quite different to a `GET` request.
```
Comment (minaorangina) [inline test/tutorial/step-7-testing-post-endpoints.md]: I'd call it the "Express workshop" or "CMS workshop", as this would be a Node Girls workshop too!

**Review by minaorangina — COMMENTED** [review]
Thank you so much for this @njseeto!! 🙏 :pray: I really appreciate the effort you've put in to this, I think a testing tutorial is a great idea.  This will be especially good for anyone who finishes the main Express workshop and wants a challenge. Could even link to it in the [extras section][] of the main tutorial.

[extras section]: https://node-girls.gitbook.io/intro-to-express/intro-to-backend-development-with-node.js/extra

I've added some suggestions for wording or explanations. Ultimately this can live in its own repo, to generate a nice gitbook. @msmichellegar @claireinez would love to hear what you think too.

File: test/readme.md (hunk 1 above)
Comment (njseeto) [reply test/readme.md]: Oh yes good shout! I've added some examples and a bit more info on why we should test

File: test/tutorial/step-1-setting-up.md (hunk 2 above)
Comment (njseeto) [reply test/tutorial/step-1-setting-up.md]: This is a good point, I have added a bit more of an explanation on scripts and more interaction at this point so students can get immediate feedback when a script is changed.

File: test/tutorial/step-1-setting-up.md (hunk 3 above)
Comment (njseeto) [reply test/tutorial/step-1-setting-up.md]: All done!

File: test/tutorial/step-2-writing-your-first-test.md (hunk 4 above)
Comment (njseeto) [reply test/tutorial/step-2-writing-your-first-test.md:11]: I decided to add a little blurb re: TDD in the main readme.
I agree, the theory of TDD might stop the flow of the coding, but it is good to know what and why TDD is important.

File: test/tutorial/step-3-setting-up-supertest.md (hunk 5 above)
Comment (njseeto) [reply test/tutorial/step-3-setting-up-supertest.md:1]: This has also been added :)

File: test/tutorial/step-3-setting-up-supertest.md (hunk 6 above)
Comment (njseeto) [reply test/tutorial/step-3-setting-up-supertest.md]: Nice pick up, it's interesting how automatic this language becomes! I have changed it so it is less jargon-y.

File: test/tutorial/step-6-asynchronous-testing.md (hunk 8 above)
Comment (njseeto) [reply test/tutorial/step-6-asynchronous-testing.md:39]: I had a little fiddle around with this and I added a line that mentions callbacks used in the workshop. I wasn't sure how in depth to go but I figure if it's mentioned, students can go back to the Express Workshop to remind themselves about callbacks and asynchronous code. Let me know what you think.

File: test/tutorial/step-7-testing-post-endpoints.md (hunk 9 above)
Comment (njseeto) [reply test/tutorial/step-7-testing-post-endpoints.md]: All done - I renamed it to "Express Workshop" 😄

File: test/tutorial/step-3-setting-up-supertest.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
+### Setting up Supertest
+
+Supertest is a library used to test HTTP requests to a Node server.t
```
Comment (minaorangina) [inline test/tutorial/step-3-setting-up-supertest.md:3]: small typo here at the end of the sentence

File: test/tutorial/step-2-writing-your-first-test.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+### Writing your first test
+Ok this is the exciting bit! We get to write our first test!!
+
+A few things before we get stuck in, Jest has 3 words that you can use:
+- `describe` : this lets you wrap a bunch of related tests together, it's used for organising test (we don't need to worry about it for now).
+- `it` : lets you run a test. We need this in every test!
+- `expect` : lets you write assertions. The tests pass if the assertions pass.
```
Comment (minaorangina) [inline test/tutorial/step-2-writing-your-first-test.md:7]: Have a sentence somewhere explaining what an assertion actually is :)

**Review by minaorangina — COMMENTED** [review]
This is looking good!  Only a couple of small nitpicks left.  I reckon this should live in its own repo, so we can generate a gitbook.  I'll set something up and you can merge in there :)

**Comment by minaorangina** [comment]
@njseeto So here's the destination repo: https://github.com/node-girls/express-testing-workshop. Clone that down and move your stuff into there.  You can use this other repo as an example for the folder structure: https://github.com/node-girls/express-workshop-steps. All being well, it _should_ just work 🤞

**Comment by njseeto** [comment]
Heya!

I've moved the tutorial into the repo and made a PR: https://github.com/node-girls/express-testing-workshop/pull/1
Have also fixed up the typo and added an explanation on assertions as suggested 😃
